### PR TITLE
feat(connector): single agent

### PR DESCRIPTION
**Pull request by sash-a**
Changed connector to be a single agent environment.

## Changes:
* Remove the `switch_perspective` method.
* Changed observation specs
* Updated tests
* Updated docs

### CODE REVIEW

**Review by dluo96 — CHANGES_REQUESTED** [review]
Thank you for your contribution @sash-a! I have left a review 😄

File: docs/environments/connector.md [hunk 1]
```diff
@@ -62,7 +49,7 @@ Agent2[path=4, position=5, target=6]
 Agent3[path=7, position=8, target=9]
 ```
 
-For example, on a 6x6 grid, the starting observation is shown below.
+This would result in a grid like this:
```
Comment (dluo96) [inline docs/environments/connector.md:52]: Isn't the previous description more informative?

File: jumanji/environments/routing/connector/env.py [hunk 2]
```diff
@@ -180,31 +177,26 @@ def step(
...
         reward = self._reward_fn(state, action, new_state)
         action_mask = jax.vmap(self._get_action_mask, (0, None))(agents, grid)
         observation = Observation(
-            grid=grids, action_mask=action_mask, step_count=new_state.step_count
+            grid=grid, action_mask=action_mask, step_count=new_state.step_count
         )
 
-        dones = jax.vmap(connected_or_blocked)(agents, action_mask)
-        discount = jnp.asarray(jnp.logical_not(dones), dtype=float)
+        dones = jnp.all(jax.vmap(connected_or_blocked)(agents, action_mask))
```
Comment (dluo96) [inline jumanji/environments/routing/connector/env.py:187]: Should this be singular?
```suggestion
        done = jnp.all(jax.vmap(connected_or_blocked)(agents, action_mask))
```

File: jumanji/environments/routing/connector/env.py [hunk 3]
```diff
@@ -180,31 +177,26 @@ def step(
...
         extras = self._get_extras(new_state)
         timestep = jax.lax.cond(
-            dones.all() | (new_state.step_count >= self.time_limit),
+            dones | (new_state.step_count >= self.time_limit),
```
Comment (dluo96) [inline jumanji/environments/routing/connector/env.py:190]: Similarly here?
```suggestion
            done | (new_state.step_count >= self.time_limit),
```

File: jumanji/environments/routing/connector/env.py [hunk 4]
```diff
@@ -180,31 +177,26 @@ def step(
...
             lambda: termination(
                 reward=reward,
                 observation=observation,
                 extras=extras,
-                shape=self.num_agents,
             ),
             lambda: transition(
                 reward=reward,
                 observation=observation,
-                discount=discount,
```
Comment (dluo96) [inline jumanji/environments/routing/connector/env.py:205]: How come the discount is removed?

File: jumanji/environments/routing/connector/env.py [hunk 5]
```diff
@@ -40,24 +40,23 @@
...
+    by allowing certain points to move to an adjacent cell at each step. However, each time a
+    point moves it leaves an impassable trail behind it. The goal is to connect all sets.
 
     - observation - `Observation`
         - action mask: jax array (bool) of shape (num_agents, 5).
         - step_count: jax array (int32) of shape ()
             the current episode step.
-        - grid: jax array (int32) of shape (num_agents, size, size)
+        - grid: jax array (int32) of shape (size, size)
             - each 2d array (size, size) along axis 0 is the agent's local observation.
```
Comment (dluo96) [inline jumanji/environments/routing/connector/env.py:60]: We have used `size` here but `grid_size` elsewhere - could we choose one and stick with it for consistency?

File: jumanji/environments/routing/connector/reward.py [hunk 6]
```diff
@@ -43,9 +43,8 @@ def __call__(
 
 
 class DenseRewardFn(RewardFn):
-    """Rewards each agent with 1.0 if it connects on that step, otherwise 0.0. Each agent also
-    receives a penalty of -0.03 at every timestep if they have not connected yet.
-    """
+    """Returns: reward of 1.0 for each agent that connects on that step and adds a penalty of
+    -0.03, per agent, at every timestep if they have not connected yet."""
```
Comment (dluo96) [inline jumanji/environments/routing/connector/reward.py:47]:
```suggestion
    """Returns: reward of 1.0 for each agent that connects on that step and adds a penalty of
    -0.03, per agent, at every timestep where they have yet to connect."""
```

File: jumanji/environments/routing/connector/reward_test.py [hunk 7]
```diff
@@ -35,27 +35,26 @@ def test_dense_reward(
...
+    assert jnp.isclose(reward, expected_reward)
 
     # Reward for some agents finished to all agents finished.
     reward = dense_reward_fn(state1, action2, state2)
-    assert (reward == jnp.array([0.0, connected_reward + timestep_reward, 0.0])).all()
+    chex.assert_rank(reward, 0)
+    assert jnp.isclose(reward, jnp.array(connected_reward + timestep_reward))
 
-    # Reward of none finished to all finished, when all agents take a non-noop action.
+    # Reward of none finished to all finished
```
Comment (dluo96) [inline jumanji/environments/routing/connector/reward_test.py:52]:
```suggestion
    # Reward for none finished to all finished
```

File: docs/environments/connector.md (hunk 1 above)
Comment (sash-a) [reply docs/environments/connector.md:52]: Technically it's not the starting observation which is why I changed it. Have reworded it better

File: docs/environments/connector.md [hunk 8]
```diff
@@ -13,26 +13,14 @@ to allow each other to connect to their own targets without overlapping.
...
 ## Observation
-At each step observation contains 3 items: a grid for each agent, an action mask for each agent and
+At each step observation contains 3 items: a grid, an action mask for each agent and
 the episode step count.
 
-- `grid`: jax array (int32) of shape `(num_agents, grid_size, grid_size)`, a 2D matrix for each
-   agent that represents pairs of points that need to be connected from the perspective of each
-   agent. The **position** of an agent has to connect to its **target**, leaving a **path** behind
+- `grid`: jax array (int32) of shape `(grid_size, grid_size)`, a 2D matrix for each
+   agent that represents pairs of points that need to be connected. Each agent has three types of
```
Comment (coyettev) [inline docs/environments/connector.md:21]: It is not "a 2D matrix for each agent" anymore, is it ?

File: jumanji/environments/routing/connector/env.py [hunk 9]
```diff
@@ -68,24 +65,22 @@ class Connector(Environment[State]):
...
-        - dense: each agent is given 1.0 if it connects on that step, otherwise 0.0. Additionally,
-            each agent that has not connected receives a penalty reward of -0.03.
+        - dense: reward is increased by 1 for each successful connection on that step. Additionally,
+            each pair of points that have not connected receives a penalty reward of -0.03.
 
     - episode termination: if an agent can't move, or the time limit is reached, or the agent
         connects to its target, it is considered done. Once all agents are done, the episode
-        terminates. The timestep discounts are of shape (num_agents,).
+        terminates. The timestep discounts are of shape (1,) and only set to `0` when all agents
+        are done.
```
Comment (coyettev) [inline jumanji/environments/routing/connector/env.py:79]: Is it still relevant to mention the discount here ? I think it is not mentioned in other single-agent envs. 
Maybe bullet points with each ending condition would be more clear for the episode termination ? like in other envs (snake or jobshop for example)

File: jumanji/environments/routing/connector/env.py [hunk 10]
```diff
@@ -180,31 +173,26 @@ def step(
             grid=grid, step_count=state.step_count + 1, agents=agents, key=state.key
         )
 
-        # Construct timestep: get observations, rewards, discounts
-        grids = self._obs_from_grid(grid)
+        # Construct timestep: get rewards, discounts
```
Comment (coyettev) [inline jumanji/environments/routing/connector/env.py:176]: Discounts are not computed here anymore, is this comment still relevant ?

File: jumanji/environments/routing/connector/env_test.py [hunk 11]
```diff
@@ -54,8 +54,8 @@ def test_connector__reset(connector: Connector, key: jax.random.KeyArray) -> Non
     assert all(is_head_on_grid(state.agents, state.grid))
     assert all(is_target_on_grid(state.agents, state.grid))
 
-    assert jnp.array_equal(timestep.discount, jnp.ones(connector.num_agents))
-    assert jnp.array_equal(timestep.reward, jnp.zeros(connector.num_agents))
+    assert jnp.array_equal(timestep.discount, jnp.asarray(1.0))
```
Comment (coyettev) [inline jumanji/environments/routing/connector/env_test.py:57]: Why not 
```python
    assert timestep.discount == 1.0
```
?

**Review by coyettev — COMMENTED** [review]
Hi Sasha, the PR looks good to me, I just left a few comments :smile:

File: docs/environments/connector.md (hunk 1 above)
Comment (dluo96) [reply docs/environments/connector.md]: Thank you!

File: jumanji/environments/routing/connector/env.py (hunk 4 above)
Comment (dluo96) [reply jumanji/environments/routing/connector/env.py:205]: Gotcha!

File: jumanji/environments/routing/connector/env.py [hunk 12]
```diff
@@ -80,11 +75,12 @@ class Connector(Environment[State]):
...
     - episode termination: if an agent can't move, or the time limit is reached, or the agent
         connects to its target, it is considered done. Once all agents are done, the episode
-        terminates. The timestep discounts are of shape (num_agents,).
+        terminates. The timestep discounts are of shape (1,) and only set to `0` when all agents
+        are done.
 
     - state: State:
         - key: jax PRNG key used to randomly spawn agents and targets.
-        - grid: jax array (int32) of shape (size, size) which corresponds to agent 0's observation.
+        - grid: jax array (int32) of shape (grid_size, grid_size) the observation.
```
Comment (dluo96) [inline jumanji/environments/routing/connector/env.py:83]:
```suggestion
        - grid: jax array (int32) of shape (grid_size, grid_size) giving the observation.
```

File: jumanji/environments/routing/connector/env.py (hunk 10 above)
Comment (sash-a) [reply jumanji/environments/routing/connector/env.py:176]: Done is computed, will change the docstring to dones instead of discount

**Review by clement-bonnet — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution.
I have 2 comments/questions:
- The registered version of Connector needs to be upped to `"Connector-v1"`.
- Have you tried running the A2C network on this? The `_make_raw_env` function in jumanji/training/setup_train.py must remove the call to `MultiToSingleWrapper` and the network probably needs to be updated to account for the new observation.

File: jumanji/environments/routing/connector/env.py [hunk 13]
```diff
@@ -68,24 +65,21 @@ class Connector(Environment[State]):
...
-            each agent that has not connected receives a penalty reward of -0.03.
+        - dense: reward is increased by 1 for each successful connection on that step. Additionally,
+            each pair of points that have not connected receives a penalty reward of -0.03.
 
-    - episode termination: if an agent can't move, or the time limit is reached, or the agent
-        connects to its target, it is considered done. Once all agents are done, the episode
-        terminates. The timestep discounts are of shape (num_agents,).
+    - episode termination:
+        * all agents either can't move (no available actions) or have connected to their target.
+        * the time limit is reached.
```
[the remark this answers is not in the document]
Comment (sash-a) [reply jumanji/environments/routing/connector/env.py:78]: It's turned into a no-op

[the remark this answers is not in the document]
Comment (clement-bonnet) [reply jumanji/environments/routing/connector/env.py:78]: Oh, yes indeed, thanks!

File: jumanji/training/configs/config.yaml [hunk 14]
```diff
@@ -1,12 +1,12 @@
 defaults:
     - _self_
-    - env: snake  # [bin_pack, cleaner, connector, cvrp, game_2048, job_shop, knapsack, maze, minesweeper, rubiks_cube, snake, tsp]
+    - env: connector  # [bin_pack, cleaner, connector, cvrp, game_2048, job_shop, knapsack, maze, minesweeper, rubiks_cube, snake, tsp]
```
Comment (clement-bonnet) [inline jumanji/training/configs/config.yaml:3]: Can we revert changes on the config before merging?

File: jumanji/training/networks/connector/actor_critic.py [hunk 15]
```diff
@@ -138,7 +132,11 @@ def __init__(
         self.env_time_limit = env_time_limit
 
     def __call__(self, observation: Observation) -> chex.Array:
-        grid = jax.vmap(process_grid)(observation.grid)  # (B, N, G, G, 2)
+        agent_ids = jnp.arange(self.num_agents)
+        # (N, B, G, G, 2)
+        grid = jax.vmap(process_grid, (None, 0))(observation.grid, agent_ids)
+        grid = jnp.transpose(grid, (1, 0, 2, 3, 4))  # (B, N, G, G, 2)
```
Comment (clement-bonnet) [inline jumanji/training/networks/connector/actor_critic.py:138]: Alternatively, you can use `out_axes` and then no need to transpose.
```suggestion
        grid = jax.vmap(process_grid, in_axes=(None, 0), out_axes=1)(observation.grid, agent_ids)
```

Comment (clement-bonnet) [inline jumanji/training/networks/connector/actor_critic.py:138]: Alternatively, you can use `out_axes` and then no need to transpose.
```suggestion
        grid = jax.vmap(process_grid, in_axes=(None, 0), out_axes=1)(observation.grid, agent_ids)  # (B, N, G, G, 2)
```

File: jumanji/environments/routing/connector/env.py [hunk 16]
```diff
@@ -391,23 +372,3 @@ def action_spec(self) -> specs.MultiDiscreteArray:
             dtype=jnp.int32,
```
Comment (clement-bonnet) [inline jumanji/environments/routing/connector/env.py:372]: In the docstring of `action_spec`, I spotted a reference to the environment being multi-agent. Could we adapt it by replacing "Since this is a multi-agent environment, the environment expects..." with "Since this is an environment with a multi-dimensional action space, it expects..." ?

**Review by clement-bonnet — DISMISSED** [review]
Thank you for your contribution. This makes the environment more consistent with the rest of the library! :rocket:

File: jumanji/environments/routing/connector/env.py [hunk 17]
```diff
@@ -70,17 +70,16 @@ class Connector(Environment[State]):
...
+        - dense: reward is 1 for each successful connection on that step. Additionally,
             each pair of points that have not connected receives a penalty reward of -0.03.
 
-    - episode termination: if an agent can't move, or the time limit is reached, or the agent
-        connects to its target, it is considered done. Once all agents are done, the episode
-        terminates. The timestep discounts are of shape (1,) and only set to `0` when all agents
-        are done.
+    - episode termination:
+        * all agents either can't move (no available actions) or have connected to their target.
+        * the time limit is reached.
```
Comment (dluo96) [inline jumanji/environments/routing/connector/env.py:78]:
```suggestion
        - all agents either can't move (no available actions) or have connected to their target.
        - the time limit is reached.
```
Do we use `*` or `-`?

File: jumanji/environments/routing/connector/env.py [hunk 18]
```diff
@@ -70,17 +70,16 @@ class Connector(Environment[State]):
...
 
     - state: State:
         - key: jax PRNG key used to randomly spawn agents and targets.
-        - grid: jax array (int32) of shape (grid_size, grid_size) the observation.
+        - grid: jax array (int32) of shape (grid_size, grid_size) giving the observation.
```
Comment (dluo96) [inline jumanji/environments/routing/connector/env.py:82]: It's not technically the observation, is it? Isn't it just a view of the connector "board"?

File: jumanji/training/networks/connector/actor_critic.py [hunk 19]
```diff
@@ -74,33 +75,35 @@ def make_actor_critic_networks_connector(
...
-        )
-        agent_channel = jnp.where(
-            agent_mask,
-            agent_grid,
-            0,
+            (agent_grid == agent_path)
+            | (agent_grid == agent_target)
+            | (agent_grid == agent_pos)
         )
+        # only current agent's info as values: 1, 2 or 3
```
Comment (dluo96) [inline jumanji/training/networks/connector/actor_critic.py:92]:
```suggestion
        # Only current agent's info as values: 1, 2 or 3
```

File: jumanji/training/networks/connector/actor_critic.py [hunk 20]
```diff
@@ -179,7 +185,7 @@ def _augment_with_step_count(
         self, embeddings: chex.Array, step_count: chex.Array
     ) -> chex.Array:
         step_count = jnp.asarray(step_count / self.env_time_limit, float)
-        num_agents = embeddings.shape[-2]
+        num_agents = self.num_agents  # embeddings.shape[-2]
```
Comment (dluo96) [inline jumanji/training/networks/connector/actor_critic.py:188]: Can we remove the commented out code?

**Review by dluo96 — DISMISSED** [review]
Left a few minor comments and suggestions

File: jumanji/environments/routing/connector/env.py (hunk 18 above)
Comment (sash-a) [reply jumanji/environments/routing/connector/env.py:82]: Now it is exactly the observation

**Review by clement-bonnet — APPROVED** [review]
